query-engine/core/src/interpreter/query_interpreters/write.rs

```diff
@@ -70,9 +70,39 @@ async fn update_one(
     q: UpdateRecord,
     trace_id: Option<String>,
 ) -> InterpretationResult<QueryResult> {
-    let res = tx.update_record(&q.model, q.record_filter, q.args, trace_id).await?;
-
-    Ok(QueryResult::Id(res))
+    let res = tx
+        .update_record(
+            q.model(),
+            q.record_filter().clone(),
+            q.args().clone(),
+            q.selected_fields(),
+            trace_id,
+        )
+        .await?;
+
+    match q {
+        UpdateRecord::WithExplicitSelection(q) => {
+            let res = res
+                .map(|res| RecordSelection {
+                    name: q.name,
+                    fields: q.selection_order,
+                    scalars: res.into(),
+                    nested: vec![],
+                    model: q.model,
+                    aggregation_rows: None,
+                })
+                .map(Box::new);
+
+            Ok(QueryResult::RecordSelection(res))
+        }
+        UpdateRecord::WithoutSelection(_) | UpdateRecord::WithImplicitSelection(_) => {
+            let res = res
+                .map(|record| record.extract_selection_result(&q.model().primary_identifier()))
+                .transpose()?;
```

Comment:
`WithoutSelection` is the old mechanism that always leads to a subsequent read (still needed for connectors that don't support returning). Therefore, we always only return `QueryResult::Id` since the result of the update won't be used for a serialized result, but only to fulfill the requirement of that subsequent read (the requirement being which id to filter on)